### PR TITLE
procedures: Dev Workspace monitoring documentation

### PR DESCRIPTION
**Pull request by dkwon17**
Signed-off-by: David Kwon <dakwon@redhat.com>
image

## What does this pull request change?
Adds Dev Workspace monitoring documentation, using Prometheus and Grafana.

## What issues does this pull request fix or reference?
https://github.com/eclipse/che/issues/20662

## Specify the version of the product this pull request applies to
This pull request applies to devworkspace-operator.

## Pull Request checklist

The author and the reviewers validate the content of this pull request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).

- Any procedure:
  - [x] Successfully tested.
- Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - Propagate the URL change in:
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/packages/dashboard-frontend/src/services/bootstrap/branding.constant.ts)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)
- [ ] Builds on [Eclipse Che hosted by Red Hat](https://workspaces.openshift.com).
- [x] the *`Validate language on files added or modified`* step reports no vale warnings.

### CODE REVIEW

File: modules/administration-guide/partials/proc_collecting-che-metrics-with-prometheus.adoc [hunk 1]
```diff
@@ -3,13 +3,13 @@
 [id="collecting-{prod-id-short}-metrics-with-prometheus_{context}"]
 = Collecting {prod-short} metrics with Prometheus
 
-This section describes how to use the Prometheus monitoring system to collect, store and query metrics about {prod-short}.
+This section describes how to use the Prometheus monitoring system to collect, store, and query metrics about {prod-short}.
```
Comment (ibuziuk) [inline modules/administration-guide/partials/proc_collecting-che-metrics-with-prometheus.adoc:6]: +1 for the the Oxford comma :+1:

File: modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+= Collecting {devworkspace} engine metrics with Prometheus
+
+[role="_abstract"]
+This section describes how to use the Prometheus to collect, store, and query metrics about the {devworkspace} engine.
+
+.Prerequisites
+
+* The `devworkspace-controller-metrics` service is exposing metrics on port `8443`.
+
+* The `devworkspace-webhookserver` service is exposing metrics on port `9443`.
```
Comment (ibuziuk) [inline modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc]: this is going to be always the case right, or is it possible to opt out from metrics exposure?

File: modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+* Prometheus 2.26.0 or later is running and the Prometheus console is running on port `9090` with a corresponding *service* and *route*. See link:https://prometheus.io/docs/introduction/first_steps/[First steps with Prometheus].
+
```
Comment (ibuziuk) [inline modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc]: would be nice to somehow introduce the section which describes the usage of the OpenShift in-cluster Prometheus for OCP. Can be done as a separate PR

**Review by ibuziuk — APPROVED** [review]
Great docs!

File: modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc (hunk 2 above)
Comment (dkwon17) [reply modules/administration-guide/partials/proc_collecting-dev-workspace-engine-metrics-with-prometheus.adoc]: I noticed that the [`devworkspace-controller-metrics`](https://github.com/devfile/devworkspace-operator/blob/main/deploy/deployment/kubernetes/objects/devworkspace-controller-metrics.Service.yaml) service is missing in some cases (ex. installation on minikube with chectl). Related issue: https://github.com/eclipse/che/issues/20762.

File: modules/administration-guide/partials/proc_viewing-dev-workspace-engine-metrics-on-grafana-dashboards.adoc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+* Prometheus is collecting metrics. See xref:proc_collecting-dev-workspace-engine-metrics-with-prometheus_{context}[].
+
+* Grafana is running on port `3000` with a corresponding *service* and *route*. See link:https://grafana.com/docs/installation/[Installing Grafana].
+
+
+.Procedure
+
+. Add the data source for the Prometheus instance.
+See .
```
Comment (themr0c) [inline modules/administration-guide/partials/proc_viewing-dev-workspace-engine-metrics-on-grafana-dashboards.adoc]: Missing xref?

File: modules/administration-guide/partials/ref_devworkspace-specific-metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,29 @@
+[id="ref_devworkspace-specific-metrics_{context}"]
+= DevWorkspace-specific metrics
+
+[role="_abstract"]
+This section describes the DevWorkspace-specific metrics exposed by the `devworkspace-controller-metrics` service.
+
+.Metrics
+* *Name:* `devworkspace_started_total` (counter) +
```
Comment (themr0c) [inline modules/administration-guide/partials/ref_devworkspace-specific-metrics.adoc]: Questionable format. Is it consistent with the rest of the docs? Should we use a definition list or a table?

File: modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc [hunk 6]
```diff
@@ -0,0 +1,35 @@
+[id="ref_grafana-dashboards-for-the-dev-workspace-engine_{context}"]
+= Grafana dashboards for the {devworkspace} engine
+
+This section describes the example Grafana dashboard (see link:https://github.com/devfile/devworkspace-operator/blob/main/doc/grafana/grafana-dashboard.json[grafana-dashboard.json]) which displays metrics collected from the {devworkspace} engine.
+
+.The *DevWorkspace-specific metrics* panel
+image::monitoring/monitoring-dev-workspace-engine-devworkspace-metrics-panel.png[Grafana dashboard panels containing metrics related to `DevWorkspace startup]
+The *DevWorkspace-specific metrics* panel contains information related to `DevWorkspace` startup.
+
+* *Average workspace start time* -- the average workspace start time
```
Comment (themr0c) [inline modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc]: Probably better with a definition list?

File: modules/administration-guide/partials/ref_devworkspace-specific-metrics.adoc (hunk 5 above)
Comment (dkwon17) [reply modules/administration-guide/partials/ref_devworkspace-specific-metrics.adoc]: Sounds great, my latest commit updates the PR to use tables:
image

File: modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc (hunk 6 above)
Comment (dkwon17) [reply modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc]: Could you provide an example of a definition list please? Is it a kind of table? I went with this format to be consistent with the existing Che-server Grafana dashboard docs: https://www.eclipse.org/che/docs/che-7/administration-guide/monitoring-che/#grafana-dashboards-for-che_che

**Comment by ibuziuk** [comment]
@themr0c please review

File: modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc (hunk 6 above)
Comment (ibuziuk) [reply modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc]: I think this is optional, let's leave it as-is for now

**Comment by nickboldt** [comment]
seems the link checker is finding a missing link:
```
========================================================================
administration-guide/monitoring-the-dev-workspace-engine/index.html
  hash does not exist --- administration-guide/monitoring-the-dev-workspace-engine/index.html --> #installation-guide:enabling-dev-workspace-engine.adoc
========================================================================
```

File: modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+.ClusterRoleBinding example
+====
+include::example$metrics-reader-cluster-role-binding.adoc[]
+====
+
+. Configure Prometheus to scrape metrics from the `8443` port exposed by the `devworkspace-controller-metrics` service and `9443` port exposed by the `devworkspace-webhookserver` service.
++
+.Prometheus configuration example
+====
+include::example$prometheus-dev-workspace-operator-config.adoc[]
```
Comment (dkwon17) [inline modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc]: My latest push rebases, renames "DevWorkspace engine" to "DevWorkspace operator", and removes `WorkspaceEngineFailure`.

Also, when building, there's an error on this line:
```
35:47  error       Did you really mean 'ato'?      CheDocs.Spelling 
```
This is part of the file name, how can I fix this?

**Comment by max-cx** [comment]
@dkwon17 Hi, David. Could you pls update your PR branch for now? Thanks

File: modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc (hunk 7 above)
Comment (max-cx) [reply modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc]: @dkwon17, questions:

1. Is there a particular reason why you opted for these blocks to be saved as separate example files? Can you include the content of all these `include`d files in this procedure module directly?

2. Is there an upstream/downstream difference in the content of the example files in this PR?

File: modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc (hunk 6 above)
Comment (themr0c) [reply modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-engine.adoc]: I apologize I used an obsolete name for it, they are now called description lists:  https://docs.asciidoctor.org/asciidoc/latest/lists/description/ 

This markup is semantically more correct than a list surcharged with additional formatting.

With a description list, the line would be:

```
Average workspace start time:: the average workspace start time
```

File: modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc (hunk 7 above)
Comment (dkwon17) [reply modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc]: Hi @max-cx ,

1. Sounds good, I will do that. There's no specific reason why these blocks were saved as separate files.
2. The examples files have no upstream/downstream difference.

**Comment by ibuziuk** [comment]
@themr0c @max-cx please, clarify what is the `CheDocs.Spelling` issue is about - Did you really mean 'ato'?

**Comment by dkwon17** [comment]
Thanks for the feedback everyone.  I have rebased and used a description list for the Grafana dashboard panel descriptions. No problems with the `CheDocs.Spelling` issue anymore since now I don't reference filenames to the yaml snippets.

**Comment by max-cx** [comment]
@dkwon17 I'll try to do a review today now that we can have the `browse built doc` preview.

File: modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc (hunk 7 above)
Comment (max-cx) [reply modules/administration-guide/partials/proc_collecting-dev-workspace-operator-metrics-with-prometheus.adoc]: @dkwon17, as far as I am aware, the examples directory is used to store snips with alternative upstream/downstream content. In my personal opinion, it's easier to just have all those blocks right inside the procedure module file for various reasons (things like the <1> reference syntax and ease of making suggested changes during review). Don't worry about resultant file length. Blocks can be either made generic using [the syntax for user-replaced values](https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values) (if it's something that the user has to enter); or, blocks can remain examples within the procedure file. So there is no need to create separate example files unless there is no way around it. Something to consider for the next time.

@themr0c, I hope I provided correct info above, if not, please correct me here.

**Comment by tstastna** [comment]
Hello @dkwon17 , I reviewed your content and created a PR with a few suggestions for changes.

**Comment by dkwon17** [comment]
Hello @tstastna , I merged the PR, thank you

**Review by tstastna — APPROVED** [review]

**Review by themr0c — APPROVED** [review]

**Comment by themr0c** [comment]
We have still broken active links:

   Non-OK status: 404 --- administration-guide/monitoring-the-dev-workspace-operator/index.html --> devfile/devworkspace-operator@main/doc/grafana/grafana-dashboard.json
  Non-OK status: 404 --- administration-guide/monitoring-the-dev-workspace-operator/index.html --> devfile/devworkspace-operator@main/doc/grafana/grafana-dashboard.json

File: modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-operator.adoc [hunk 8]
```diff
@@ -0,0 +1,34 @@
+[id="ref_grafana-dashboards-for-the-dev-workspace-operator_{context}"]
+= Grafana dashboards for the {devworkspace} operator
+
+This section describes the example Grafana dashboard, see link:https://github.com/devfile/devworkspace-operator/blob/main/doc/grafana/grafana-dashboard.json[grafana-dashboard.json], which displays metrics collected from the {devworkspace} operator.
```
Comment (themr0c) [inline modules/administration-guide/partials/ref_grafana-dashboards-for-the-dev-workspace-operator.adoc]:
```suggestion
This section describes the example Grafana dashboard, see link:https://github.com/devfile/devworkspace-operator/blob/main/docs/grafana/grafana-dashboard.json[grafana-dashboard.json], which displays metrics collected from the {devworkspace} operator.
```

File: modules/administration-guide/partials/proc_viewing-dev-workspace-operator-metrics-on-grafana-dashboards.adoc [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+* Grafana is running on port `3000` with a corresponding *service* and *route*. See link:https://grafana.com/docs/installation/[Installing Grafana].
+
+
+.Procedure
+
+. Add the data source for the Prometheus instance.
+See link:https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-data-source[Creating a Prometheus data source].
+
+. Import the example link:https://github.com/devfile/devworkspace-operator/blob/main/doc/grafana/grafana-dashboard.json[grafana-dashboard.json] dashboard.
```
Comment (themr0c) [inline modules/administration-guide/partials/proc_viewing-dev-workspace-operator-metrics-on-grafana-dashboards.adoc]:
```suggestion
. Import the example link:https://github.com/devfile/devworkspace-operator/blob/main/docs/grafana/grafana-dashboard.json[grafana-dashboard.json] dashboard.
```

**Comment by themr0c** [comment]
Fixed the broken links (typo)

**Pull request merged**